File: Searching/Searching.cpp
#include <bits/stdc++.h>

using namespace std;

class Searching {
public:
    int interpolationSearch(int a[], int n, int target) {    
        int lo = 0, hi = (n - 1);
    
        while (lo <= hi && target >= a[lo] && target <= a[hi]) {
            if (lo == hi) {
                if (a[lo] == target) return lo;
                return -1;
            }

            int pos = lo + (((double)(hi - lo) / (a[hi] - a[lo])) * (target - a[lo]));
    
            if (a[pos] == target)
                return pos;
    
            if (a[pos] < target)
                lo = pos + 1;
    
            else
                hi = pos - 1;
        }
        return -1;
    }

    int binarySearchIterative(int a[], int left, int right, int target) {
        while (left <= right) {
            int mid = left + (right - left) / 2;
    
            if (a[mid] == target) {
                return mid;
            }

            if (a[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }

    int binarySearchRecursion(int a[], int left, int right, int target) {
        if (right >= left) {
            int mid = left + (right - left) / 2;
    
            if (a[mid] == target) {
                return mid;
            }

            if (a[mid] > target) {
                return binarySearchRecursion(a, left, mid - 1, target);
            }

            return binarySearchRecursion(a, mid + 1, right, target);
        }
    
        return -1;
    }
};

int main() {
    return 0;
}

File: Searching/Searching.py
from typing import List

class Searching():
    # Unordered/ Ordered list: O(n)
    def linear_search(self, arr: List[int], target: int) -> int:
        for i in range(len(arr)):
            if arr[i] == target:
                return i
        return -1
    
    # Ordered list: Divide and Conquer O(logn) 
    def binary_search_recursive(self, arr: List[int], left:int, right:int, target: int) -> int:
        if left <= right:
            mid = (left + right) // 2

            if arr[mid] == target:
                return mid
            elif arr[mid] < target:
                return self.binary_search_recursive(arr, mid+1, right, target)
            else: 
                return self.binary_search_recursive(arr, left, mid-1, target)
        else:
            return -1
    
    def binary_search(self, arr: List[int], target: int) -> int:
        if sorted(arr) != arr:
            arr = sorted(arr)

        left = 0
        right = len(arr) - 1

        return self.binary_search_recursive(arr, left, right, target)


if __name__ == '__main__':
    searchAlg = Searching()
    
    arr = [4,5,6,7,0,1,2]

    print(searchAlg.linear_search(arr, 0))

    sorted_arr = [1, 3, 5, 6, 7, 8]
    print(searchAlg.binary_search(sorted_arr, 5))